### ISSUE TITLE
Include documentation from an external file

### ISSUE THREAD

**Issue by AlanBlanchet** [microsoft/TypeScript#51330]
Hello vscode community !

I have a feature that I've been thinking about for a long time now. Whenever I try to document my code I want to be super efficient and not make my codebase too big. 

## Introduction

Let's take a quick example to introduce the issue. I will be using `javascript`.

```javascript
/**
 * Adds `a` and `b` together
 */
function sum(a, b) {
    return a + b;
}

/**
 * Adds `a` and `b` together
 */
function addition(a, b) {
    return sum(a,b)
}
```

Of course theres no logic to the code but please treat it as if there was a logic.

## Feature

The feature I would like to add is the including of documentation from an external file like below.
Here is the file structure : 
```
|
| -- index.js
| -- sumDoc.txt
```

In `index.js` I have the following : 
```javascript
/**
 * {@include ./sumDoc.txt}
 */
function sum(a, b) {
    return a + b;
}

/**
 * {@include ./sumDoc.txt}
 */
function addition(a, b) {
    return sum(a,b)
}
```

In `sumDoc.txt` I have the following : 
```
Adds `a` and `b` together
```

Of course when I hover the mouse over one of the two functions, `sum` and `addition`, I would like vscode to show me the documentation like it would show up in the very first example.
![image](https://user-images.githubusercontent.com/22567306/198252563-75802f90-38d0-4b26-8e96-072543c9cb2e.png)

## Why ?

If this feature was to be in production, 

- It would totally respect the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle. Even if this principle mostly applies to programming paradigms in general, it would also perfectly suit documentation. When you change the documentation for multiple functions, you only need to change it in the `txt` file...
- When you have a long documentation in your codebase, you tend to get lost in your file and overuse your mouses scroll wheel.
- You could use these `txt` files seperatly from the codebase and use them elsewhere. For example inside a custom website.

## External features

- Adding a `.doc` extension for files. Because currently you can't get the nice linting when using the `@` prefix from vscode when creating a documentation file.
- Why not make the including of documentation a one line thing (instead of 3 in my examples)
  ```javascript
  // @include ./sumDoc.txt
  function sum(a, b) {
      return a + b;
  }
  
  // @include ./sumDoc.txt
  function addition(a, b) {
      return sum(a,b)
  }
  ```
  Maybe this feature is included by default in vscode ?

Thank you in advance for considering this feature !

**Comment by fatcerberus** [comment]
Personally I would find the potentially thousands of `.txt` files and not being able to read the documentation of a function without cross-referencing an external file to not be worth the tradeoff of shorter code.  DRY isn’t a factor here because it seems like a very rare case where you’d be able to use the same file for different functions.  Especially since JSDoc is more than just a plain-English description and also usually includes parameter types, return type, etc.

FWIW if all you have is a description and don’t care about documenting parameters, etc., then you can already do that on one line today:

```ts
/** This function does a thing */
function doTheThing() { doIt(); }
```

**Comment by RyanCavanaugh** [comment]
This seems like a very big hammer for a rather rare nail. How often do you have functions with identical descriptions?

**Comment by AlanBlanchet** [comment]
> This seems like a very big hammer for a rather rare nail. How often do you have functions with identical descriptions? Doesn't this become a refactoring hazard, since if you update `someDoc.txt` when changing the behavior of `addition`, you might be invalidating the description of `sum`?

Yes I agree with you. But that's the job of the developer to make sure not to invalidate the description of other functions. Just like in OOP when you change a parent class you have to make sure not to break anything on the classes that inherit from it. I don't really have identical descriptions for the hole functions. But I do for parts of descriptions. But I see what you mean.

> Personally I would find the potentially thousands of `.txt` files and not being able to read the documentation of a function without cross-referencing an external file to not be worth the tradeoff of shorter code. DRY isn’t a factor here because it seems like a very rare case where you’d be able to use the same file for different functions. Especially since JSDoc is more than just a plain-English description and also usually includes parameter types, return type, etc.
> 
> FWIW if all you have is a description and don’t care about documenting parameters, etc., then you can already do that on one line today:
> 
> ```ts
> /** This function does a thing */
> function doTheThing() { doIt(); }
> ```

I'm sure it could be useful ! I'll show you some other use cases : 

### Function parameters

Currently in some of my projects I have functions which take the same parameter but do something different with it. Thus I have a duplicate description of this parameter in all my functions. There's currently no other way than to copy/paste this description. And whenever you update the comment for that parameter, you have to remember that it is also on another function.

### Examples

Examples is a big part to me. Instead of having 200 lines of examples in the same file as the function, you could put them in a separate file. Of course in the editor you can close the 200 lines of comment to make it 1 line. But still...

### Real life examples where it would be useful

- In the [pandas python](https://pandas.pydata.org/docs/index.html) lib, there are huge amounts of documentations. Which is really good. Whenever in vscode you hover over the class/functions/variables... They nearly always show pretty comments with useful information. When you look at all the docs you can clearly see that in some areas there are lots of redundant descriptions. (this also works for `numpy` I believe)
- I'm an experienced `react` developer. If I take for example the famous `useState()` hook, all the doc is online. The comment only serves as a bridge between the user in vscode and the react online doc since theres a @link tag inside redirecting to the official react doc. Maybe it's overkill to say "Include the doc in vscode so you don't have to click on the link to go to the website. I'm too lazy". And I agree. But wouldn't it bring a better dev experience to just hover over the element and get all of the docs without having to access an external source ? This isn't really easy to do for the react devs currently I think.

I was also thinking about something else I missed. Would it be possible to get this kind of `.txt` file from a server ? I say `.txt` but of course as I said there surely are better alternatives. (like a `.doc` extension)
```javascript
/*
 * {@include {http://mydocumentation.com/a.txt}}
 */
function a() {}
```

Although I do understand and recognize that it might be too much work for what you get out of it when you say.
> This seems like a very big hammer for a rather rare nail.

or

> not be worth the tradeoff of shorter code

To me it would help in documenting my code. And others too I hope.

Thank you for your feedbacks

**Comment by fatcerberus** [comment]
I don't think people are regularly putting 200+ lines of example code in the JSDoc for a function, and even if they are that's not something you're going to want the IDE to show in a hover description, so this suggested `@include` feature doesn't seem useful for that.

These files almost certainly won't be directly reusable anyway.  Rarely do you just have a description and the JSDoc block more often looks something like this:

```ts
/**
 * Add `a` and `b` together.
 * @param {number} a - first number to add
 * @param {number} b - second number to add
 */
function add(a, b) { return a + b; }
```

There are a lot of different JSDoc tags, and it's *highly* unlikely you'd have a large number of functions with these all being completely identical.  So you'd end up with a confusing multi-line block containing several `@include` directives, and at that point the benefits re: code size become much less compelling.

IMO comments are for humans first and only incidentally for machine use, so concerns like DRY don't really apply.  But that's just me.

**Comment by AlanBlanchet** [comment]
Here is an image I took for one of my projects in Python. As you can see, the description of the `replace` function is pretty long (re. The scrollbar on the right).
![image](https://user-images.githubusercontent.com/22567306/198587119-9877f6fa-f745-4026-a706-23ca66edd478.png)

It might be because it's a really popular function but still. It has comments included when hovering and I find it to be really useful.

Also, in some editors like [Spyder](https://www.spyder-ide.org/) you can get this kind of doc by pressing `F1` on a class/function/variable. (it might be connected to an API though)
Examples take the most space in the documentation. At least for how they do it in Python. There aren't much examples for `JSDoc`. It would be cool though. I don't count anymore how often I visit the [MDN](https://developer.mozilla.org/en-US/) to find what I'm looking for.

> So you'd end up with a confusing multi-line block containing several @include directives, and at that point the benefits re: code size become much less compelling.

I don't really see why it would become confusing. I'll take a short example. When you have 2 variables `a` and `b` that each have their own documentation. Their documentation can be as long as how explicit you believe you have to make it so devs can understand easily. The code you would have to create in order to do this is the following : 
```javascript
/**
 * Add `a` and `b` together.
 * @include {./a.txt}
 * @include {./b.txt}
 */
function add(a, b) { return a + b; }
```

Does that really seam confusing ? What can be confusing though is that you can start having circular references...
Which I guess is kind of hard and will get you a headache to solve...
